### PR TITLE
(PUP-9395) specify version-specific documentation params with install

### PR DESCRIPTION
**Pull request by tkishel**
With this commit, install specifies version-specific parameters to not install
documentation, and specify those parameters independent of the source parameter.

### CODE REVIEW

File: lib/puppet/provider/package/gem.rb [hunk 1]
```diff
@@ -124,13 +130,20 @@ def install(useversion = true)
           command << source
         else
           # interpret it as a gem repository
-          command << "--source" << "#{source}" << resource[:name]
+          command << "--source" << "#{source}"
```
Comment (joshcooper) [inline lib/puppet/provider/package/gem.rb]: Copying my comment from the other PR. This change looks unrelated.

File: lib/puppet/provider/package/gem.rb [hunk 2]
```diff
@@ -92,9 +92,15 @@ def insync?(is)
     is.any? { |version| dependency.match?('', version) }
   end
 
+  def gem_version(cmd)
+    command = [cmd, '--version']
+    execute(command, {:failonfail => true, :combine => true, :custom_environment => {"HOME"=>ENV["HOME"]}})
```
Comment (joshcooper) [inline lib/puppet/provider/package/gem.rb]: IIRC, the `gem` command is sensitive to `HOME` due to `.gemrc`, and this matches the behavior below. The tricky part is if you override any options to `execute`, then you have to also respecify the defaults for `failonfail` and `combine`. It might be good to have a general `gem_command` method that does the option handling, and appends whatever options are passed to it, eg `--version` in this case or the array of arguments on line 147.

File: lib/puppet/provider/package/gem.rb [hunk 3]
```diff
@@ -124,13 +130,20 @@ def install(useversion = true)
...
         end
       end
+    end
+
+    command_version = gem_version(command(:gemcmd))
+    if Puppet::Util::Package.versioncmp(command_version, '2.0.0') == -1
+      command << '--no-rdoc' << '--no-ri'
     else
-      command << "--no-rdoc" << "--no-ri" << resource[:name]
+      command << '--no-document'
```
Comment (joshcooper) [inline lib/puppet/provider/package/gem.rb]: Rubygems added [--no-document in 2.0](https://github.com/rubygems/rubygems/commit/0d22d6025ce2653ec9fb20a4e0745c79bda677da). We could probably simplify this to always using `--no-document` provided we bump the [minimum rubygems version to 2.0 in puppet's .gemspec](https://github.com/puppetlabs/puppet/blob/5.5.x/.gemspec#L20).

File: lib/puppet/provider/package/gem.rb (hunk 1 above)
Comment (tkishel) [reply lib/puppet/provider/package/gem.rb]: If the purpose of specifying —no-rdoc  etc is to avoid generating/installing gem documentation (and avoid the time necessary to do so), then it is a logical error to conditionally not do so based upon the souce parameter. 

I’ll remove this change tomorrow,  double check my work and submit the correction as part of another PR.

File: lib/puppet/provider/package/gem.rb (hunk 2 above)
Comment (tkishel) [reply lib/puppet/provider/package/gem.rb]: I think I follow, I’ll try that tomorrow.

File: lib/puppet/provider/package/gem.rb (hunk 3 above)
Comment (tkishel) [reply lib/puppet/provider/package/gem.rb]: Wouldn’t that introduce errors when managing gems in system rubies that ship with rubygems older than 2.0 and be a breaking change? I recognize that the potential is low ...

Comment (tkishel) [reply lib/puppet/provider/package/gem.rb]: For example:

```
[root@pe-agent ~]# cat /etc/redhat-release 
CentOS release 6.6 (Final)

[root@pe-agent ~]# yum install ruby rubygems -y -q

[root@pe-agent ~]# gem install colorize --no-document
ERROR:  While executing gem ... (OptionParser::InvalidOption)
    invalid option: --no-document
```

File: lib/puppet/provider/package/gem.rb (hunk 1 above)
Comment (tkishel) [reply lib/puppet/provider/package/gem.rb]: Ah, Charlie helped me understand the issue (with just `<< resource[:name]`) ... will adjust.

File: lib/puppet/provider/package/gem.rb (hunk 3 above)
Comment (joshcooper) [reply lib/puppet/provider/package/gem.rb]: Yeah, never mind about my comment. I forgot that the `gem` command we're executing is not necessarily the same version as the ruby version running puppet.

**Comment by tkishel** [comment]
Added a commit implementing a common execute_gem_command method.
This commit also reverts the unrelated change with source.

File: lib/puppet/provider/package/gem.rb [hunk 4]
```diff
@@ -112,35 +132,35 @@ def install(useversion = true)
...
+          command_options << source
         else
           # interpret it as a gem repository
-          command << "--source" << "#{source}" << resource[:name]
+          command_options << "--source" << "#{source}" << resource[:name]
         end
       end
     else
-      command << "--no-rdoc" << "--no-ri" << resource[:name]
+      command_options << resource[:name]
```
Comment (joshcooper) [inline lib/puppet/provider/package/gem.rb]: Previously we only appended `--no-rdoc` & `--no-ri` when `resource[:source]` was `nil`. Is it safe to always append them as is done in this PR? I tested a simple case that seems ok, but what about the others?

```
$ gem install ./colorize-0.8.1.gem --no-document
Successfully installed colorize-0.8.1
1 gem installed
```

Comment (tkishel) [reply lib/puppet/provider/package/gem.rb]: The issue you pointed out, and Charlie confirmed, was that we only want to add `resource[:name]` (the name of the gem package) when `:source` is specified and we interpret `:source` as a gem repository. 

I did some research and found nothing about the `--no-` documentation flags being incompatible with the other cases, and here are tests:

```
# gem repository
[root@pe-master ~]# gem install --no-document --source https://rubygems.org colorize 
Successfully installed colorize-0.8.1
1 gem installed

# absolute file path
[root@pe--master ~]# gem install --no-document /root/colorize-0.8.1.gem 
Successfully installed colorize-0.8.1
1 gem installed
```

File: lib/puppet/provider/package/gem.rb (hunk 1 above)
Comment (tkishel) [reply lib/puppet/provider/package/gem.rb]: Reverted the unrelated change involving `:source` in the first commit.

File: lib/puppet/provider/package/gem.rb (hunk 2 above)
Comment (tkishel) [reply lib/puppet/provider/package/gem.rb]: Added a second commit implementing a general `execute_gem_command` method.

**Comment by tkishel** [comment]
Both commits pass testing, so the second commit could be deleted if we do not want its general gem command method.

**Comment by joshcooper** [comment]
jenkins please test this

**Comment by tkishel** [comment]
I found an issue:

`CommandDefiner` in provider.rb creates convenience execution methods that set `failonfail`, `combine`, and optionally, `environment`. When a child provider defines its own command via `commands` or `has_command`, the provider-specific path is always returned by `command`. But when the convenience execution method is invoked, the last convenience method to be defined is executed. This makes invoking those convenience execution methods unsuitable for inherited providers.

In this case, causing the puppet_gem provider to inherit the parent gem provider's convenience `gemcmd` methods, with the wrong path. So we cannot use `has_command` to set `failonfail`, `combine`, and optionally, `environment` in one location. 

I will replace the last commit with one that defines a class method `execute_gem_command` to do the same.

**Comment by branan** [comment]
@tkishel it looks like this PR has two commits in it with the same message - should they be squashed? Otherwise I think we're good to get this merged now

**Comment by tkishel** [comment]
That's because Josh asked for:

> It might be good to have a general `gem_command` method

So if we did want that, we could squash them;
and if not, we could drop the second commit and just update the document flags :)

Can you squash them, or do I need to?

**Comment by branan** [comment]
@tkishel Ah, I see what's in the second commit now. I think that's enough of a change from what this PR does that it should be its own commit still, just with a new commit message describing how stuff got moved to the common method (and why). Once you've got that reworded, I think I can merge this :+1:

**Comment by tkishel** [comment]
So ...

Modify the body of the second commit message and/or the title?
And keep the second commit in this PR?

**Comment by branan** [comment]
Yeah, keep the second command and update both the title and body (I guess mark it as `(maint)` in the title, since it's not really part of fixing PUP-9395?)
